### PR TITLE
Provide parser with path to script file executed by means of -File command line argument

### LINKED ISSUE #4217: Unix: A script without .ps1 extension passed to the powershell binary lacks invocation information ($PSCommandPath, $MyInvocation), such as when invoked via a shebang line.
Steps to reproduce
------------------

Create file `./t` - note the absence of extension `.ps1` - with the content below and make it executable (`chmod +x ./t`).

```powershell
#!/usr/bin/env powershell

'$PSCommandPath: ' + $PSCommandPath
'$MyInvocation.MyCommand.Path: ' + $MyInvocation.MyCommand.Path
'$MyInvocation: ' + ($MyInvocation | Out-String)
```

Both the following invocation methods, from `bash`, yield the behavior described below:

```sh
./t  # invocation via shebang line
powershell ./t   # implied -File
```

Expected behavior
-----------------

`$PSCommandPath` and `$MyInvocation.MyCommand.Path` should reflect the script's file path, and `$MyInvocation` should be populated appropriately.

Actual behavior
---------------

```none
$PSCommandPath:
$MyInvocation.MyCommand.Path:
$MyInvocation:

MyCommand             : #!/usr/bin/env powershell

                        '$PSCommandPath: ' + $PSCommandPath
                        '$MyInvocation.MyCommand.Path: ' + $MyInvocation.MyCommand.Path
                        '$MyInvocation: ' + ($MyInvocation | Out-String)

BoundParameters       : {}
UnboundArguments      : {}
ScriptLineNumber      : 0
OffsetInLine          : 0
HistoryId             : 1
ScriptName            :
Line                  :
PositionMessage       :
PSScriptRoot          :
PSCommandPath         :
InvocationName        :
PipelineLength        : 2
PipelinePosition      : 1
ExpectingInput        : False
CommandOrigin         : Runspace
DisplayScriptPosition :
```

Environment data
----------------

```powershell
PowerShell Core v6.0.0-beta.3 on macOS 10.12.5
PowerShell Core v6.0.0-beta.3 on Ubuntu 16.04.1 LTS
```

### PR DESCRIPTION
**Pull request by m5x**
# PR Summary

Allow `System.Management.Automation.Runspaces.Command` operating in `IsScript = true` mode to remember and pass on path to the script file and make use of this feature when executing script file passed as -File command line argument.

## PR Context

Fixes #4217

### CODE REVIEW

**Review by vexx32 — COMMENTED** [review]
Code seems sound, nice work!

Can we add a test to verify this works and guard against regressions?

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 1]
```diff
@@ -73,6 +73,12 @@ internal Command(string command, bool isScript, bool? useLocalScope)
             _useLocalScope = useLocalScope;
         }
 
+        internal Command(string script, string scriptPath, bool? useLocalScope)
+            : this(script, true, useLocalScope)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:77]:
```suggestion
            : this(script, isScript: true, useLocalScope)
```

**Review by iSazonov — COMMENTED** [review]
Please add tests.

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 1 above)
Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:78]: I didn't do this because of the convention used in this class - [the first constructor](https://github.com/PowerShell/PowerShell/blob/16ec1cc32f9e60c9d76b074c865c32a5d2acf5a2/src/System.Management.Automation/engine/hostifaces/Command.cs#L28) in the class also doesn't use named `isScript` parameter. Do you want to add this change anyway?

**Comment by m5x** [comment]
Tests have been added. I also added one small change to use `ExternalScriptInfo` for scripts with known path (i.e. scripts started with -File argument) and thus provide proper values in $MyInvocation special variable even for scripts without .ps1 extension.

**Comment by m5x** [comment]
Sorry for the spacing around parentheses. I like to put spaces inside for better readability and I've been doing it so long that I type it automatically this way and sometimes forget to check and fix it when committing to other projects. I will fix it in next commit - should I add [Feature] prefix to the commit message again to comply with  [testing guidelines](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr) in case it is the last one of this PR?

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 1 above)
Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:78]: Yes, it makes code more readable.

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 2]
```diff
@@ -497,6 +509,16 @@ CommandOrigin origin
                     commandProcessorBase = new CommandProcessor(functionInfo, executionContext,
                                                                 _useLocalScope ?? false, fromScriptFile: false, sessionState: executionContext.EngineSessionState);
                 }
+                else if (ScriptPath != null)
+                {
+                    var scriptName = Path.GetFileName(ScriptPath);
+                    var scriptInfo = new ExternalScriptInfo( scriptName, ScriptPath, executionContext);
+                    commandProcessorBase = new DlrScriptCommandProcessor( scriptInfo, 
+                        executionContext,_useLocalScope ?? false,
+                        executionContext.EngineSessionState);
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:518]: Formatting:
```suggestion
                    commandProcessorBase = new DlrScriptCommandProcessor(
                        scriptInfo, 
                        executionContext,
                        _useLocalScope ?? false,
                        executionContext.EngineSessionState);
```

Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:518]: This is the same kind of case like above - I did it first the way you suggest but then I reformatted it to keep the convention already used - [see this call](https://github.com/PowerShell/PowerShell/blob/16ec1cc32f9e60c9d76b074c865c32a5d2acf5a2/src/System.Management.Automation/engine/hostifaces/Command.cs#L502-L506) a few lines below. 

So the general approach is to ignore conventions already used and use formatting as suggested by CodeFactor analyzer? In that case should I also change the comment for the ScriptPath property to start with word 'Get' even though comments for other properties in that class start with word 'Access'?

Comment (vexx32) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:518]: @m5x yeah there's some older code in a fair few places that we're more or less trying to update as we come across it. In general prefer the newer syntax.

As for the doc comments, I'd think we should go with the CodeFactor recommendation there. Since they're only comments, too, is it worth updating the other non-compliant documentation comments in this file do you think @iSazonov?

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:518]: It is not strong request - we could fix formatting later. But the PR is small and the code was already changed so we could fix formatting now.

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 1 above)
Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs]: Fixed with last commit together with other occurrences in the same class.

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 2 above)
Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs]: Fixed with last commit. Also fixed the property docs in this class which did start with word 'access' instead of 'gets'.

**Comment by m5x** [comment]
Added the requested formatting and property docs changes. CodeFactor reports no more issues.

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 3]
```diff
@@ -497,13 +509,27 @@ CommandOrigin origin
...
-                    commandProcessorBase = new DlrScriptCommandProcessor(scriptBlock,
-                                                                         executionContext, _useLocalScope ?? false,
-                                                                         origin,
-                                                                         executionContext.EngineSessionState,
-                                                                         DollarUnderbar);
+                    commandProcessorBase = new DlrScriptCommandProcessor(
+                        scriptBlock,
+                        executionContext,
+                        _useLocalScope ?? false,
+                        origin,
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:530]: Why do we ignore this above if `ScriptPath != null`?

Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: I am not sure I follow. If we have path to the script we can use the same `DlrScriptCommandProcessor` constructor as would be used if the file had .ps1 extension. What is being ignored in your opinion?

Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: The constructor that gets passed `ExternalScriptInfo` as the first argument is necessary for the `InvocationInfo` (`$MyInvocation` special var) to be created properly.

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: My ask is about `origin` argument.

Comment (m5x) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: I guess the answer to this question is the same as answer to question why `DlrScriptCommandProcessor` constructor for external script does not accept `origin` parameter? In my opinion it's because it is not need at this stage where we already know the command is external script.

Note that the `if` above for cases when `scriptBlock.UsesCmdletBinding` is `true` also does not pass the `origin` to the `CommandProcessor` constructor so it's the same case in that regard. Isn't it because `CommandOrigin` is needed only in special cases and during command searching/discovery?

I have seen PowerShell's source code for the first time a few days ago so I definitely do not feel competent enough to answer such questions but I guess that when it gets looked at by somebody who knows the inner workings really well he/she will be immediately able to tell whether it's secure to do it this way. In my opinion it should be because the code path that gets executed when the script file has the `.ps1` extension does exactly the same (in addition to command discovery which is redundant in our case).

Comment (iSazonov) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: Thanks! I did not look in depth but I'd still expect that `origin` should be here.

**Comment by iSazonov** [comment]
@m5x Please rebase to pass CIs.

**Comment by m5x** [comment]
Is rebasing all that's blocking merge of this pull request? I have currently some other things I have to finish first but if I am the only one blocking this I'll try to find a time to rebase this weekend.

**Comment by iSazonov** [comment]
It is a question for MSFT team - @daxian-dbw @anmenaga have you time to review the PR?

**Comment by daxian-dbw** [comment]
Add @SteveL-MSFT and @rjmholt to review.
@SteveL-MSFT You marked the issue "by design" in https://github.com/PowerShell/PowerShell/issues/4217#issuecomment-662134332, can you please review this PR and the subsequent comments in that issue to see if we want to accept this PR? Also would the change of design cause any breaking change?

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 4]
```diff
@@ -25,7 +26,7 @@ public sealed class Command
         /// <param name="command">Name of the command or script contents.</param>
         /// <exception cref="ArgumentNullException">Command is null.</exception>
         public Command(string command)
-            : this(command, false, null)
+            : this(command, isScript: false, null)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:29]: Worth labelling the `null` parameter here too

**Review by rjmholt — APPROVED** [review]

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 4 above)
Comment (rjmholt) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:29]:
```suggestion
            : this(command, isScript: false, useLocalScope: null)
```

File: src/System.Management.Automation/engine/hostifaces/Command.cs (hunk 3 above)
Comment (rjmholt) [reply src/System.Management.Automation/engine/hostifaces/Command.cs:530]: Just looked into this. I'll say that the constructor code here is old and frankly hard to follow (could use some style improvements).

But when the `CommandInfo` is provided, the origin is taken to be `Internal`:

https://github.com/PowerShell/PowerShell/blob/1223518a437ccc06c49357358150919db9b91eab/src/System.Management.Automation/engine/ScriptCommandProcessor.cs#L38

**Comment by rjmholt** [comment]
Blocked on https://github.com/PowerShell/PowerShell/pull/12494#issuecomment-668136216; @SteveL-MSFT or the @PowerShell/powershell-committee need to advise on whether we want to make this change

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 5]
```diff
@@ -497,13 +509,27 @@ CommandOrigin origin
                     commandProcessorBase = new CommandProcessor(functionInfo, executionContext,
                                                                 _useLocalScope ?? false, fromScriptFile: false, sessionState: executionContext.EngineSessionState);
                 }
+                else if (ScriptPath != null)
```
Comment (iSazonov) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:512]:
```suggestion
                else if (ScriptPath is not null)
```

**Review by jazzdelightsme — APPROVED** [review]
:shipit:

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 6]
```diff
@@ -257,6 +257,39 @@ Describe "ConsoleHost unit tests" -tags "Feature" {
...
         }
+
+        It "MyInvocation.MyCommand.Name should contain script file name when executing script with -File argument" -TestCases @(
+            @{Filename = "test.ps1"},
+            @{Filename = "test-no-ext"}
+        ) {
+            param($Filename)
+            $testFilePath = Join-Path $testdrive $Filename
+            Set-Content -Path $testFilePath -Value '$MyInvocation.MyCommand.Name'
+            $observed = & $powershell -File $testFilePath -NoProfile -NoLogo
```
Comment (daxian-dbw) [inline test/powershell/Host/ConsoleHost.Tests.ps1:268]: There was [an intentional breaking change](https://github.com/PowerShell/PowerShell/pull/15859) in PowerShell 7.2 related to `powershell -File`: when running on Windows, a file without `.ps1` extension is not allowed to run using `pwsh -File`. So, the test won't work on Windows.

File: src/System.Management.Automation/engine/hostifaces/Command.cs [hunk 7]
```diff
@@ -497,13 +509,27 @@ CommandOrigin origin
...
+                {
+                    var scriptName = Path.GetFileName(ScriptPath);
```
Comment (daxian-dbw) [inline src/System.Management.Automation/engine/hostifaces/Command.cs:514]: The script (content of the file) could be using cmdlet binding (`scriptBlock.UsesCmdletBinding` being `true`), by having a param block defined. In that case, the code execution would fall into the `if (scriptBlock.UsesCmdletBinding)` block, right?

**Review by daxian-dbw — CHANGES_REQUESTED** [review]
